File: src/www/__test__/interpreter/InstructionSequence.js
import chalk from "chalk";

export default class InstructionSequence {
  constructor(instructions) {
    this._instructions = instructions;
    this._currentIndex = 0;
  }

  hasNext() {
    return this._currentIndex < this._instructions.length;
  }

  next() {
    const currentInstruction = this._instructions[this._currentIndex];
    this._currentIndex += 1;
    return currentInstruction;
  }

  toString() {
    const instructions = this._instructions;
    const count = instructions.length;
    const previousIndex = this._currentIndex - 1;
    const result = [];

    let index = 0;
    while (index < previousIndex && index < count) {
      result.push(`  ${this._format(index, instructions, chalk.green.dim)}`);
      index += 1;
    }

    if (index < count) {
      result.push(`> ${this._format(index, instructions, chalk.red.bold)}`);
      index += 1;
    } else {
      result.push(`> EOS`);
    }

    while (index < count) {
      result.push(`  ${this._format(index, instructions, chalk.yellow.dim)}`);
      index += 1;
    }

    return result.join("\n");
  }

  _format(index, instructions, color = x => x) {
    const { line, source } = instructions[index];
    return color(`${`    ${line}`.slice(-5)}: ${source}`);
  }
}

File: src/test/java/com/drpicox/game/interpreter/InstructionSequence.java
package com.drpicox.game.interpreter;

import java.util.ArrayList;
import java.util.List;

class InstructionSequence {
    private List<Instruction> instructions;
    private int currentIndex;

    public InstructionSequence(List<Instruction> instructions) {
        this.instructions = new ArrayList<>(instructions);
        this.currentIndex = 0;
    }

    public boolean hasNext() {
        return currentIndex < instructions.size();
    }

    public Instruction next() {
        var currentInstruction = instructions.get(currentIndex);
        currentIndex += 1;
        return currentInstruction;
    }

    public String getPreviousSource() {
        var previousIndex = currentIndex - 1;
        if (previousIndex < 0) return "null";

        return instructions.get(previousIndex).getSource();
    }

    public String getPreviousLine() {
        var previousIndex = currentIndex - 1;
        if (previousIndex < 0) return "-1";

        return instructions.get(previousIndex).getLine();
    }

    @Override
    public String toString() {
        var size = instructions.size();
        var previousIndex = currentIndex - 1;
        var result = new StringBuilder();

        var index = 0;
        while (index < previousIndex && index < size) {
            result.append("   √ ").append(instructions.get(index)).append('\n');
            index += 1;
        }

        if (index < size) {
            result.append(">  X ").append(instructions.get(index)).append('\n');
            index += 1;
        } else {
            result.append("> EOS\n");
        }

        while (index < size) {
            result.append("   · ").append(instructions.get(index)).append('\n');
            index += 1;
        }

        return result.toString();
    }

    public void deplete() {
        currentIndex = instructions.size() + 1;
    }
}
